diff --color -urN --no-dereference vulkan-tools-clean/cube/CMakeLists.txt vulkan-tools-workdir/cube/CMakeLists.txt
--- vulkan-tools-clean/cube/CMakeLists.txt	2025-09-22 18:14:43.000000000 +0300
+++ vulkan-tools-workdir/cube/CMakeLists.txt	2025-09-25 01:53:02.372996170 +0300
@@ -69,7 +69,7 @@
 
 option(BUILD_WSI_DISPLAY_SUPPORT "Build DISPLAY WSI support" ${WSI_DISPLAY_DEFAULT_SETTING})
 
-if (CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU")
+if (CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU|FishOS")
     option(BUILD_WSI_XCB_SUPPORT "Build XCB WSI support" ON)
     option(BUILD_WSI_XLIB_SUPPORT "Build Xlib WSI support" ON)
     option(BUILD_WSI_WAYLAND_SUPPORT "Build Wayland WSI support" ON)
@@ -173,7 +173,7 @@
     list(APPEND ENABLED_CUBE_PLATFORMS VK_USE_PLATFORM_ANDROID_KHR)
 elseif(APPLE)
     list(APPEND ENABLED_CUBE_PLATFORMS VK_USE_PLATFORM_METAL_EXT)
-elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU")
+elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU|FishOS")
     if(BUILD_WSI_XCB_SUPPORT)
         list(APPEND ENABLED_CUBE_PLATFORMS VK_USE_PLATFORM_XCB_KHR)
     endif()
@@ -224,7 +224,7 @@
 # a missing dependency on libpthread. While newer Ubuntu versions use a glibc version where libpthread
 # is integrated into libc, older ubuntu's do not so we need to link threads directly in order for
 # validation layers to be loadable.
-if (CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU")
+if (CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU|FishOS")
     find_package(Threads REQUIRED)
 endif()
 
@@ -240,7 +240,7 @@
 
     add_subdirectory(android)
 
-elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU")
+elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU|FishOS")
     add_executable(vkcube)
     target_sources(vkcube PRIVATE
         cube.c
@@ -312,7 +312,7 @@
 
 if(APPLE)
     add_subdirectory(macOS/cubepp)
-elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU")
+elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU|FishOS")
     add_executable(vkcubepp
                    cube.cpp
                    ${PROJECT_SOURCE_DIR}/cube/cube.vert
diff --color -urN --no-dereference vulkan-tools-clean/cube/cube.c vulkan-tools-workdir/cube/cube.c
--- vulkan-tools-clean/cube/cube.c	2025-09-22 18:14:43.000000000 +0300
+++ vulkan-tools-workdir/cube/cube.c	2025-09-25 01:52:38.727159906 +0300
@@ -5087,7 +5087,7 @@
         }
     }
 }
-#elif defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__QNX__) || defined(__GNU__)
+#elif defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__QNX__) || defined(__GNU__) || defined(__fishos__)
 int main(int argc, char **argv) {
     struct demo demo;
 
diff --color -urN --no-dereference vulkan-tools-clean/cube/cube.cpp vulkan-tools-workdir/cube/cube.cpp
--- vulkan-tools-clean/cube/cube.cpp	2025-09-22 18:14:43.000000000 +0300
+++ vulkan-tools-workdir/cube/cube.cpp	2025-09-25 01:52:36.241765092 +0300
@@ -4133,7 +4133,7 @@
 }
 
 #elif defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__QNX__) || defined(__GNU__) || \
-    defined(__Fuchsia__)
+    defined(__Fuchsia__) || defined(__fishos__)
 
 template <WsiPlatform WSI_PLATFORM>
 void Demo::execute() {
diff --color -urN --no-dereference vulkan-tools-clean/cube/gettime.h vulkan-tools-workdir/cube/gettime.h
--- vulkan-tools-clean/cube/gettime.h	2025-09-22 18:14:43.000000000 +0300
+++ vulkan-tools-workdir/cube/gettime.h	2025-09-25 01:52:33.205197573 +0300
@@ -26,7 +26,7 @@
 
 #include <windows.h>
 
-#elif defined(__unix__) || defined(__linux) || defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__ANDROID__) || defined(__EPOC32__) || defined(__QNX__)
+#elif defined(__unix__) || defined(__linux) || defined(__linux__) || defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__ANDROID__) || defined(__EPOC32__) || defined(__QNX__) || defined(__fishos__)
 
 #include <time.h>
 
@@ -55,7 +55,7 @@
         return count.QuadPart / (freq.QuadPart / ns_in_s);
     }
 
-#elif defined(__unix__) || defined(__linux) || defined(__linux__) || defined(__ANDROID__) || defined(__QNX__)
+#elif defined(__unix__) || defined(__linux) || defined(__linux__) || defined(__ANDROID__) || defined(__QNX__) || defined(__fishos__)
     struct timespec currTime;
     clock_gettime(CLOCK_MONOTONIC, &currTime);
     return (uint64_t)currTime.tv_sec * ns_in_s + (uint64_t)currTime.tv_nsec;
diff --color -urN --no-dereference vulkan-tools-clean/icd/CMakeLists.txt vulkan-tools-workdir/icd/CMakeLists.txt
--- vulkan-tools-clean/icd/CMakeLists.txt	2025-09-22 18:14:43.000000000 +0300
+++ vulkan-tools-workdir/icd/CMakeLists.txt	2025-09-25 01:53:02.372996170 +0300
@@ -38,7 +38,7 @@
     endif()
 elseif(BUILD_MOCK_ANDROID_SUPPORT)
     add_definitions(-DVK_USE_PLATFORM_ANDROID_KHR)
-elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU")
+elseif(CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU|FishOS")
     if(BUILD_WSI_XCB_SUPPORT)
         add_definitions(-DVK_USE_PLATFORM_XCB_KHR -DVK_USE_PLATFORM_XCB_KHX)
     endif()
diff --color -urN --no-dereference vulkan-tools-clean/vulkaninfo/CMakeLists.txt vulkan-tools-workdir/vulkaninfo/CMakeLists.txt
--- vulkan-tools-clean/vulkaninfo/CMakeLists.txt	2025-09-22 18:14:43.000000000 +0300
+++ vulkan-tools-workdir/vulkaninfo/CMakeLists.txt	2025-09-25 01:53:02.374996157 +0300
@@ -62,7 +62,7 @@
 
 target_compile_definitions(vulkaninfo PRIVATE VK_ENABLE_BETA_EXTENSIONS VK_NO_PROTOTYPES)
 
-if (CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU")
+if (CMAKE_SYSTEM_NAME MATCHES "Linux|BSD|GNU|FishOS")
     option(BUILD_WSI_XCB_SUPPORT "Build XCB WSI support" ON)
     option(BUILD_WSI_XLIB_SUPPORT "Build Xlib WSI support" ON)
     option(BUILD_WSI_WAYLAND_SUPPORT "Build Wayland WSI support" ON)
