diff --color -urN --no-dereference neofetch-clean/neofetch neofetch-workdir/neofetch
--- neofetch-clean/neofetch	2025-10-05 14:15:02.354104389 +0300
+++ neofetch-workdir/neofetch	2025-10-05 14:15:04.897108265 +0300
@@ -929,6 +929,7 @@
         AIX)      os=AIX ;;
         IRIX*)    os=IRIX ;;
         FreeMiNT) os=FreeMiNT ;;
+        FishOS)   os=FishOS ;;
 
         Linux|GNU*)
             os=Linux
@@ -954,6 +955,7 @@
     [[ $distro ]] && return
 
     case $os in
+        FishOS) distro=FishOS ;;
         Linux|BSD|MINIX)
             if [[ -f /bedrock/etc/bedrock-release && $PATH == */bedrock/cross/* ]]; then
                 case $distro_shorthand in
